### PR TITLE
Add v2 audit logs for action management component

### PR DESCRIPTION
**Pull request by Shenali-SJ**
### Description
This PR adds v2 audit logs for the action management component. Logs are recorded for create, update, delete processes, and the activation/deactivation of actions.

**Related Issue:**
- https://github.com/wso2/product-is/issues/21261

---

### Sample Audit Logs:

ADD
```
{
   "id":"<id>",
   "recordedAt":"2024-10-09T17:37:55.437605Z",
   "requestId":"<requested-id>",
   "initiatorId":"<initiator-id>",
   "initiatorType":"User",
   "targetId":"System",
   "targetType":"Action",
   "action":"add-action",
   "data":{
      "ActionName":"test",
      "EndpointConfiguration":{
         "EndpointUri":"https://wso2is.free.beeceptor.com",
         "AccessToken":"w************************************e",
         "EndpointAuthenticationScheme":"BEARER"
      }
   }
}
```

UPDATE
(Action name update)
```
{
   "id":"<id>",
   "recordedAt":"2024-10-09T18:33:26.329017Z",
   "requestId":"<requested-id>",
   "initiatorId":"<initiator-id>",
   "initiatorType":"User",
   "targetId":"System",
   "targetType":"Action",
   "action":"update-action",
   "data":{
      "ActionName":"test21",
      "ActionId":"e0a2b0da-77d0-486d-a6f8-f97bdbfd8a05"
   }
}
```

ACTIVATE
```
{
   "id":"<id>",
   "recordedAt":"2024-10-09T17:41:39.044124Z",
   "requestId":"<requested-id>",
   "initiatorId":"initiator-id>",
   "initiatorType":"User",
   "targetId":"System",
   "targetType":"Action",
   "action":"activate-action",
   "data":{
      "ActionId":"e0a2b0da-77d0-486d-a6f8-f97bdbfd8a05",
      "ActionType":"preIssueAccessToken"
   }
}
```

### CODE REVIEW

**Comment by malithie** [comment]
@Shenali-SJ even for the data let's follow the camel case. Or existing logs use pascal case for the keys in the json object representing data ?

File: components/action-mgt/org.wso2.carbon.identity.action.management/pom.xml [hunk 1]
```diff
@@ -96,13 +100,20 @@
...
                             org.wso2.carbon.identity.secret.mgt.core.exception; version="${carbon.identity.package.import.version.range}",
                             org.wso2.carbon.identity.secret.mgt.core.model; version="${carbon.identity.package.import.version.range}",
                             org.wso2.carbon.utils; version="${carbon.kernel.package.import.version.range}",
+                            org.wso2.carbon.identity.event.*; version="${carbon.identity.package.import.version.range}",
+                            org.wso2.carbon.identity.central.log.mgt.utils; version="${carbon.identity.package.import.version.range}",
+                            org.wso2.carbon; version="${carbon.kernel.package.import.version.range}",
+                            org.wso2.carbon.context; version="${carbon.kernel.package.import.version.range}",
+                            org.wso2.carbon.utils.*; version="${carbon.kernel.package.import.version.range}",
+                            org.wso2.carbon.user.core.*; version="${carbon.kernel.package.import.version.range}"
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/pom.xml:116]: I don't see this package used

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java [hunk 2]
```diff
@@ -38,6 +41,12 @@
 import java.util.Map;
 import java.util.UUID;
 
+import static org.wso2.carbon.identity.central.log.mgt.utils.LogConstants.ActionManagement.ACTIVATE_ACTION;
+import static org.wso2.carbon.identity.central.log.mgt.utils.LogConstants.ActionManagement.ADD_ACTION;
+import static org.wso2.carbon.identity.central.log.mgt.utils.LogConstants.ActionManagement.DEACTIVATE_ACTION;
+import static org.wso2.carbon.identity.central.log.mgt.utils.LogConstants.ActionManagement.DELETE_ACTION;
+import static org.wso2.carbon.identity.central.log.mgt.utils.LogConstants.ActionManagement.UPDATE_ACTION;
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java:48]: IMO, we don't need to add these constants to the central log mgt unit. These are related to the component. So the component can define them

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java [hunk 3]
```diff
@@ -78,6 +88,10 @@ public Action addAction(String actionType, Action action, String tenantDomain) t
         validateMaxActionsPerType(resolvedActionType, tenantDomain);
         doPreAddActionValidations(action);
         String generatedActionId = UUID.randomUUID().toString();
+        if (LoggerUtils.isEnableV2AuditLogs()) {
+            JSONObject data = auditLogger.generateAuditLogEntry(ADD_ACTION, actionType, null, action);
+            auditLogger.printAuditLog(ADD_ACTION, data);
+        }
         return CACHE_BACKED_DAO.addAction(resolvedActionType, generatedActionId, action,
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java:95]: So the real update happens here. That means audit log should be done after this

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java [hunk 4]
```diff
@@ -123,6 +137,10 @@ public Action updateAction(String actionType, String actionId, Action action, St
         String resolvedActionType = getActionTypeFromPath(actionType);
         Action existingAction = checkIfActionExists(resolvedActionType, actionId, tenantDomain);
         doPreUpdateActionValidations(action);
+        if (LoggerUtils.isEnableV2AuditLogs()) {
+            JSONObject data = auditLogger.generateAuditLogEntry(UPDATE_ACTION, actionType, actionId, action);
+            auditLogger.printAuditLog(UPDATE_ACTION, data);
+        }
         return CACHE_BACKED_DAO.updateAction(resolvedActionType, actionId, action, existingAction,
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java:144]: Same comment above applies here

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtV2AuditLoggerUtil.java [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+import static org.wso2.carbon.identity.central.log.mgt.utils.LogConstants.ActionManagement.PASSWORD_FIELD;
+import static org.wso2.carbon.identity.central.log.mgt.utils.LogConstants.ActionManagement.UPDATE_ACTION;
+import static org.wso2.carbon.identity.central.log.mgt.utils.LogConstants.ActionManagement.USERNAME_FIELD;
+import static org.wso2.carbon.identity.central.log.mgt.utils.LoggerUtils.jsonObjectToMap;
+import static org.wso2.carbon.identity.central.log.mgt.utils.LoggerUtils.triggerAuditLogEvent;
+
+/**
+ * Action management V2 audit logger util class.
+ */
+public class ActionMgtV2AuditLoggerUtil {
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtV2AuditLoggerUtil.java:57]: ActionManagementAuditLogger

Comment (malithie) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtV2AuditLoggerUtil.java:57]: IMO, Using v1 of cetral logger or v2 is an implementation decision of this class. So let's keep it that way

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtV2AuditLoggerUtil.java [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+                data = createAuditLogsForUpdateAction(actionId, action);
+                break;
+            case DELETE_ACTION:
+            case ACTIVATE_ACTION:
+            case DEACTIVATE_ACTION:
+                data = generateAuditLogsById(actionType, actionId);
+                break;
+        }
+        return data;
+    }
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtV2AuditLoggerUtil.java:124]: Do we really need to make this public. We can still use this from a single log method IMO.

**Comment by Shenali-SJ** [comment]
> @Shenali-SJ even for the data let's follow the camel case. Or existing logs use pascal case for the keys in the json object representing data ?

We are using the Pascal case in existing logs. Therefore, the same case will be used here to preserve the consistency.

File: components/action-mgt/org.wso2.carbon.identity.action.management/pom.xml (hunk 1 above)
Comment (Shenali-SJ) [reply components/action-mgt/org.wso2.carbon.identity.action.management/pom.xml:116]: These packages are being used in the [ActionMgtAuditLoggerUtil](https://github.com/wso2/carbon-identity-framework/blob/a105238b1c38f7405ecf43ecaa9c07b9b0de7b7f/components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtAuditLoggerUtil.java) class to retrieve current user data and initiator information.

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java (hunk 2 above)
Comment (Shenali-SJ) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java]: Refactored with a105238b1c38f7405ecf43ecaa9c07b9b0de7b7f

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java (hunk 3 above)
Comment (Shenali-SJ) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java]: Updated with a105238b1c38f7405ecf43ecaa9c07b9b0de7b7f

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java (hunk 4 above)
Comment (Shenali-SJ) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java]: Updated with a105238b1c38f7405ecf43ecaa9c07b9b0de7b7f

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtV2AuditLoggerUtil.java (hunk 5 above)
Comment (Shenali-SJ) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtV2AuditLoggerUtil.java]: Refactored with a105238b1c38f7405ecf43ecaa9c07b9b0de7b7f

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java [hunk 7]
```diff
@@ -48,6 +49,7 @@ public class ActionManagementServiceImpl implements ActionManagementService {
     private static final CacheBackedActionMgtDAO CACHE_BACKED_DAO =
             new CacheBackedActionMgtDAO(new ActionManagementDAOImpl());
     private static final ActionValidator ACTION_VALIDATOR = new ActionValidator();
+    private static final ActionMgtAuditLoggerUtil auditLogger = new ActionMgtAuditLoggerUtil();
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java:52]: Shall we say ActionManagementAuditLogger. So it's more consistent with naming used now

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtAuditLoggerUtil.java [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+import org.wso2.carbon.utils.AuditLog;
+import org.wso2.carbon.utils.multitenancy.MultitenantUtils;
+
+import static org.wso2.carbon.identity.central.log.mgt.utils.LoggerUtils.jsonObjectToMap;
+import static org.wso2.carbon.identity.central.log.mgt.utils.LoggerUtils.triggerAuditLogEvent;
+
+/**
+ * Action management V2 audit logger util class.
+ */
+public class ActionMgtAuditLoggerUtil {
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtAuditLoggerUtil.java:40]: Gave a suggestion for the name above

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtAuditLoggerUtil.java [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+
+    /**
+     * To get the current user, who is doing the current task.
+     *
+     * @return current logged-in user.
+     */
+    private static String getUser() {
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtAuditLoggerUtil.java:47]: We don't need to have static methods. The whole point of instantiating is to easily unit test and you can't do that with static methods

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtAuditLoggerUtil.java [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+     * @param action    Action object to be logged.
+     */
+    public void printActionAuditLog(Operation operation, String actionId, Action action) {
+
+        if (!LoggerUtils.isEnableV2AuditLogs()) {
+            return;
+        }
+        JSONObject data = createAuditLogEntry(actionId, action);
+        buildAuditLog(operation, data);
+
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtAuditLoggerUtil.java:111]: Remove this line

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtAuditLoggerUtil.java [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+    private static class LogConstants {
+
+        public static final String ACTION_TYPE_FIELD = "ActionType";
+        public static final String ACTION_NAME_FIELD = "ActionName";
+        public static final String ACTION_ID_FIELD = "ActionId";
+        public static final String ACTION_DESCRIPTION_FIELD = "ActionDescription";
+        public static final String ACTION_STATUS_FIELD = "ActionStatus";
+        public static final String ENDPOINT_CONFIG_FIELD = "EndpointConfiguration";
+        public static final String ENDPOINT_URI_FIELD = "EndpointUri";
+        public static final String ENDPOINT_AUTHENTICATION_SCHEME_FIELD = "EndpointAuthenticationScheme";
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtAuditLoggerUtil.java:315]: Shall we just say AuthenticationScheme ? This is anyway within EndpointConfiguration json object

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtAuditLoggerUtil.java [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+     * Create audit log data with action and ID.
+     * This method expects null/empty action fields.
+     *
+     * @param actionId ID of the action to be logged.
+     * @param action   Action to be logged.
+     * @return audit log data.
+     */
+    private JSONObject createAuditLogEntry(String actionId, Action action) {
+
+        JSONObject data = new JSONObject();
```
Comment (Thisara-Welmilla) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtAuditLoggerUtil.java:176]: In this method, we are checking values are null and if not we add to the data. Why we need to have another createAuditLogEntry(Action action). IMO,  we can utilize the same method.

Comment (Shenali-SJ) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtAuditLoggerUtil.java:176]: `createAuditLogEntry(Action action) ` this method is expected to be used with ADD operation where each field should be non-null/non-empty. `createAuditLogEntry(String actionId, Action action)` this method is to be used with UPDATE operations where we expect null/empty fields by design.

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java (hunk 7 above)
Comment (Shenali-SJ) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java]: This is updated with b5355995d46dbafa80bb825fa0da60e5bd5062ee

Comment (Shenali-SJ) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java]: This is updated with b5355995d46dbafa80bb825fa0da60e5bd5062ee

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtAuditLoggerUtil.java (hunk 8 above)
Comment (Shenali-SJ) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtAuditLoggerUtil.java]: This is updated with b5355995d46dbafa80bb825fa0da60e5bd5062ee

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtAuditLoggerUtil.java (hunk 9 above)
Comment (Shenali-SJ) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtAuditLoggerUtil.java]: This is updated with b5355995d46dbafa80bb825fa0da60e5bd5062ee

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtAuditLoggerUtil.java (hunk 10 above)
Comment (Shenali-SJ) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtAuditLoggerUtil.java]: Fixed with b5355995d46dbafa80bb825fa0da60e5bd5062ee

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtAuditLoggerUtil.java (hunk 11 above)
Comment (Shenali-SJ) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtAuditLoggerUtil.java]: This is updated with b5355995d46dbafa80bb825fa0da60e5bd5062ee

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java [hunk 13]
```diff
@@ -48,6 +49,7 @@ public class ActionManagementServiceImpl implements ActionManagementService {
     private static final CacheBackedActionMgtDAO CACHE_BACKED_DAO =
             new CacheBackedActionMgtDAO(new ActionManagementDAOImpl());
     private static final ActionValidator ACTION_VALIDATOR = new ActionValidator();
+    private static final ActionManagementAuditLoggerUtil auditLogger = new ActionManagementAuditLoggerUtil();
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java:52]: Shall we remove 'Util' from class name

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java [hunk 14]
```diff
@@ -78,8 +80,10 @@ public Action addAction(String actionType, Action action, String tenantDomain) t
         validateMaxActionsPerType(resolvedActionType, tenantDomain);
         doPreAddActionValidations(action);
         String generatedActionId = UUID.randomUUID().toString();
-        return CACHE_BACKED_DAO.addAction(resolvedActionType, generatedActionId, action,
+        Action createdAction = CACHE_BACKED_DAO.addAction(resolvedActionType, generatedActionId, action,
                 IdentityTenantUtil.getTenantId(tenantDomain));
+        auditLogger.printActionAuditLog(ActionManagementAuditLoggerUtil.Operation.ADD, createdAction);
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java:85]: shall we just say "printAuditLog"

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java (hunk 13 above)
Comment (Shenali-SJ) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java]: This is updated with c81491d0d97828a7760ca1c4dc028f0ab9cd0b8b

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java (hunk 14 above)
Comment (Shenali-SJ) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionManagementServiceImpl.java]: This is updated with c81491d0d97828a7760ca1c4dc028f0ab9cd0b8b

**Review by malithie — APPROVED** [review]

**Review by malithie — APPROVED** [review]

**Pull request merged**